File: 257. Binary Tree Paths.java
//找到所有树上的路径，采用深度优先的方法，没有使用栈来模拟，因为递归的方法可以免去改变String在父亲节点的值。  
  public static ArrayList<String> arr;
    public static void tree(TreeNode node, String s)
    {
        if(node.left == null && node.right == null)
        {
            arr.add(new String(s + node.val));
        }
        else if(node.left == null)
        {
            tree(node.right, new String(s + node.val + "->"));
        }
        else if(node.right == null)
        {
            tree(node.left, new String(s + node.val + "->"));
        }
        else
        {
            tree(node.left, new String(s + node.val + "->"));
            tree(node.right, new String(s + node.val + "->"));
        }
    }
   
    public List<String> binaryTreePaths(TreeNode root)
    {
        arr = new ArrayList<String>();
        StringBuffer sb = new StringBuffer();
        if(null == root)
            return arr;
        tree(root, "");
        return arr;
    }

File: 257. Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:

    vector<string> binaryTreePaths(TreeNode* root)
    {
        if(!root)
            return ret;
        rr = root;
        dfs(root);
        return ret;
    }
private:
    vector<string> ret;
    string now;
    TreeNode *rr;
    void dfs(TreeNode *node)
    {
        TreeNode *r = node->right;
        TreeNode *l = node->left;
        string now_bak = now;
        if(node == rr)
            now += to_string(node->val);
        else
            now += "->" + to_string(node->val);
        if(!r && !l)
            ret.push_back(now);
        else if(!r)
            dfs(l);
        else if(!l)
            dfs(r);
        else
        {
            dfs(r);
            dfs(l);
        }
        now = now_bak;
    }
};